bundles/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/handler/ICloudAccountBridgeHandler.java

```diff
@@ -88,85 +101,281 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+            if (!service.isTrustedSession()) {
+                service.trustSession();
+            }
+            if (!service.isTrustedSession()) {
+                System.err.println("Trust failed!!!");
+            }
+     * </pre>
+     *
+     * The call to {@link ICloudService#authenticate(boolean)} request a token from the user.
+     * This should be done only once. Afterwards the user has to update the configuration.
+     * In openhab this method here is called for several reason (e.g. config change). So we track if we already
+     * requested a code {@link #validate2faCode}.
+     */
+    private boolean checkLogin() {
+
+        logger.debug("Starting iCloud authentication (AuthState={}, Thing={})...", authState,
+                getThing().getUID().getAsString());
+        if (authState == AuthState.WAIT_FOR_CODE) {
+            throw new IllegalStateException("2-FA authentication not completed.");
+        }
+
         try {
-            logger.debug("iCloud bridge starting handler ...");
             ICloudAccountThingConfiguration config = getConfigAs(ICloudAccountThingConfiguration.class);
-            final String localAppleId = config.appleId;
-            final String localPassword = config.password;
-            if (localAppleId != null && localPassword != null) {
-                connection = new ICloudConnection(localAppleId, localPassword);
-            } else {
+            boolean success = false;
+
+            // No code requested yet or session is trusted (hopefully).
+            success = this.iCloudService.authenticate(false);
+            if (!success) {
+                authState = AuthState.USER_PW_INVALID;
+                logger.warn("iCloud authentication failed. Invalid credentials.");
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Invalid credentials.");
+                this.iCloudService = null;
+                return false;
+            }
+            if (iCloudService.requires2fa()) {
+                // New code was requested. Wait for the user to update config.
+                logger.warn(
+                        "iCloud authentication requires 2-FA code. Please provide code configuration for thing '{}'.",
+                        getThing().getUID().getAsString());
+                authState = AuthState.WAIT_FOR_CODE;
                 updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                        "Apple ID/Password is not set!");
-                return;
+                        "Please provide 2-FA code in thing configuration.");
+                return false;
+            }
+
+            if (!this.iCloudService.isTrustedSession()) {
+                logger.debug("Trying to establish session trust.");
+                success = this.iCloudService.trustSession();
+                if (!success) {
+                    logger.warn("iCloud trust session failed.");
```

Comment:
This logging is not needed as the framework already logs the status change below.